app/domain/authentication/authn_k8s/execute_command_in_container.rb

```diff
@@ -0,0 +1,141 @@
+        env:                           ENV,
+        websocket_client:              WebSocket::Client::Simple,
+        ws_client_event_handler_class: WebSocketClientEventHandler,
+        message_log_class:             MessageLog,
+        validate_message:              MessageLog::ValidateMessage.new,
+        logger:                        Rails.logger
+      },
+      inputs:       %i(k8s_object_lookup pod_namespace pod_name container cmds body stdin)
+    ) do
+
+      extend Forwardable
+      def_delegators :@k8s_object_lookup, :kube_client
+
+      DEFAULT_KUBE_EXEC_COMMAND_TIMEOUT = 5
+
+      def call
+        init_ws_client
+        add_websocket_event_handlers
+        wait_for_close_message
+        verify_channel_is_closed
+        verify_error_stream_is_empty
+        websocket_messages
+      end
+
+      private
+
+      def init_ws_client
+        ws_client
+        ws_client_event_handler
+      end
+
+      def ws_client
+        @ws_client ||= @websocket_client.connect(server_url, headers: headers)
+      end
+
+      def ws_client_event_handler
+        @ws_client_event_handler ||= @ws_client_event_handler_class.new(
+          ws_client:         ws_client,
+          stdin:             @stdin,
+          body:              @body,
+          pod_name:          @pod_name,
+          validate_message:  @validate_message,
+          message_log_class: @message_log_class,
+          logger:            @logger
+        )
+      end
+
+      def add_websocket_event_handlers
+        # We need to set this as if we'll use @ws_client_event_handler inside
+        # the curly brackets it will look for such a member in the websocket_client
+        # class
+        ws_client_event_handler = @ws_client_event_handler
+
+        ws_client.on(:open) { ws_client_event_handler.on_open }
+        ws_client.on(:message) { |msg| ws_client_event_handler.on_message(msg) }
+        ws_client.on(:close) { ws_client_event_handler.on_close }
+        ws_client.on(:error) { |err| ws_client_event_handler.on_error(err) }
+      end
+
+      def wait_for_close_message
+        (timeout / 0.1).to_i.times do
+          break if ws_client_event_handler.channel_closed
+          sleep 0.1
+        end
+      end
+
+      def verify_channel_is_closed
+        return if ws_client_event_handler.channel_closed
+
+        raise Errors::Authentication::AuthnK8s::ExecCommandTimedOut.new(
+          timeout,
+          @container,
+          @pod_name
+        )
+      end
+
+      def verify_error_stream_is_empty
+        error_stream = ws_client_event_handler.message_log.messages[:error]
+        return if error_stream.nil? || error_stream.empty?
+        raise Errors::Authentication::AuthnK8s::ExecCommandError.new(
+          @container,
+          @pod_name,
+          websocket_error(error_stream)
+        )
+      end
+
+      def server_url
+        api_uri  = kube_client.api_endpoint
+        base_url = "wss://#{api_uri.host}:#{api_uri.port}"
```

Comment:
Take the values from configuration?

Reply:
why not from the kubeclient? why add more configuration?

Reply:
If this pattern can change it should be configurable.